include/mpscringbuffer.hpp

```diff
@@ -0,0 +1,52 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+namespace ddprof {
+//  mimic: std::hardware_destructive_interference_size, C++17
+constexpr std::size_t hardware_destructive_interference_size = 128;
+
+struct MPSCRingBufferMetaDataPage {
+  alignas(hardware_destructive_interference_size) uint64_t writer_pos;
+  alignas(hardware_destructive_interference_size) uint64_t reader_pos;
+  alignas(hardware_destructive_interference_size) std::atomic_flag spinlock;
+};
+
+inline void lock(std::atomic_flag *lock) {
+  static constexpr uint32_t k_max_active_spin = 4000;
+  static constexpr std::chrono::nanoseconds k_yield_sleep =
+      std::chrono::microseconds(500);
+
+  uint32_t spincount = 0;
+
+  for (;;) {
+    if (!lock->test_and_set(std::memory_order_acquire)) {
```

Comment:
should we consider a second safety mechanism to get out ?

Reply:
I see @r1viollet's point:  if a thread/process dies while having the lock, then other processes will wait endlessly.
I will add a timeout.